File: frontend/src/api/history.ts
import { apiRequest } from './client'

export interface RecentPlay {
  spotify_id: string
  name: string
  artist_names: string[]
  album_name: string
  album_image_url: string | null
  duration_ms: number
  spotify_url: string | null
  played_at: string
}

export interface RecentHistoryResponse {
  limit: number
  items: RecentPlay[]
}

export function getRecentHistory(
  limit = 50,
): Promise<RecentHistoryResponse> {
  const query = new URLSearchParams({
    limit: String(limit),
  })

  return apiRequest<RecentHistoryResponse>(
    `/api/history/recent?${query.toString()}`,
  )
}

File: backend/app/schemas/history.py
from datetime import datetime

from pydantic import BaseModel, Field


class RecentPlayResponse(BaseModel):
    spotify_id: str
    name: str
    artist_names: list[str] = Field(
        default_factory=list,
    )
    album_name: str
    album_image_url: str | None = None
    duration_ms: int
    spotify_url: str | None = None
    played_at: datetime


class RecentHistoryResponse(BaseModel):
    limit: int
    items: list[RecentPlayResponse] = Field(
        default_factory=list,
    )